Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,176 @@
+Unlike the above `coreclr` based APIs these don't require the caller to fully specify all startup parameters, instead these APIs understand artifacts produced by .NET Core SDK making it much easier to consume SDK produced apps/libraries.
+The native app is still required to locate the `hostfxr` or `hostpolicy` libraries. These APIs are also designed for specific narrow scenarios, any usage outside of these bounds is typically not possible.
+
+
+## Scope
+This document focuses on easy-to-use hosting which cooperates with the .NET Core SDK and consumes the artifacts produced by building the managed app/libraries directly. It completely ignores the COM-style ABI as it's hard to use from some programming languages.
+
+As such the document explicitly excludes any hosting based on directly loading `coreclr`. The document focuses on using the existing .NET Core hosting layer in new ways. For details on the .NET Core hosting components see [this document](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md).
+
+
+## High-level proposal
+In .NET Core 3.0 the hosting layer (see [here](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md)) ships with several hosts. These are binaries which act as the entry points to the .NET Core hosting/runtime:
+* The "muxer" (`dotnet.exe`)
+* The `apphost` (`.exe` which is part of the app)
+* The `comhost` (`.dll` which is part of the app and acts as COM server)
+* The `ijwhost` (`.dll` consumed via `.lib` used by IJW assemblies)
+
+Every one of these hosts serve different scenario and expose different APIs. The one thing they have in common is that their main purpose is to find the right `hostfxr`, load it and call into it to execute the desired scenario. For the most part all these hosts are basically just wrappers around functionality provided by `hostfxr`.
+
+The proposal is to add a new host library `nethost` which can be used by native apps to easily host managed components and to easily locate `hostfxr` for more advanced scenarios.
+
+At the same time add the ability to pass additional runtime properties when starting the runtime through the hosting entry points (starting app, loading component). This can be used by the native app to:
+* Register startup hook without modifying environment variables (which are inherited by child processes)
+* Introduce new runtime knobs which are only available for native hosts without the need to update the hosting APIs every time.
+
+This new ability would be added to both `hostfxr` and to the `nethost`.
+
+
+## New host binary for component hosting
+Add new library `nethost` which will act as the easy to use host for loading managed components.
```

Comment:
Can AppHost be implemented as one customization of nethost (for code/scenario reuse)? For now, there's only framework-dependent apps with nethost, bit this is not a fundamental technical limitation, right?

Reply:
Re framework dependent: Self-contained `nethost` while possible introduces interesting problems - for one it would only allow loading one component. Trying to load a second component into that process would become really tricky, since we could not rely on framework resolution to validate/match framework dependencies. Instead this resolution would have to happen on assembly level which is obviously much more complicated. So while possible, I would like to see a real world scenario for this first.

Re `apphost` == `nethost`: All the hosts are essentially the same: `apphost`, `comhost`, `ijwhost` and now `nethost` all do basically the same thing - locate the `hostfxr` and call into it. They differ in two things:
* What do they call in `hostfxr`
* How do they expose the functionality

In that regard if we were to unify anything, then `comhost` and `nethost` are actually very similar. Also `ijwhost` is similar to them to a degree. They all allow loading "components" - isolated islands of managed code (they will use fully isolated ALC to load the code into). They generate TPA which only contain frameworks - and they only support framework dependent apps.

`apphost` is the outlier here. 
* It's an `exe` unlike all the others which are `dlls`. I know that on Windows it's possible for an `exe` to be loaded as a `dll` and ask for exports, so in that sense it's possible to have both. I don't know about Linux/macOS - if it's even possible to have such a binary.
* It runs the app - which effectively means it "owns" the process. For example returning from `Main` will terminate the process. Unlike all the other hosts which don't have that power.
* It loads the managed code into the default ALC, meaning it produces TPA which contains everything, frameworks and app.

Then on Windows there's the added complexity of allowing customization of `apphost` (native resources, name and so on).

From a testing point of view it's kind of unfortunate since we need to test that many binaries, but from a shipping point of view it kind of doesn't matter. Each needs its own support in SDK, so there's no real gain from sharing the binary. Not counting that `nethost` is special in this sense as it won't be integrated into SDK - it will be its own NuGet package (or something like that).

Reply:
>  Not counting that nethost is special in this sense as it won't be integrated into SDK - it will be its own NuGet package (or something like that).

We need nethost in every app that we are trying to run using AspNetCoreModule, having export on apphost would be much more preferable as it would allow us to avoid having two copies of the same functionality in every customers app.

Reply:
I actually don't think you need nethost in every app - you need nethost in your code - to find hostfxr… has very little to do with the app. nethost has to be 100% backward compatible anyway, so it should work on old apps as well.

If we add an export to apphost it would only work for new apps, it won't fix your problem with existing apps.

And more importantly nethost to find hostfxr doesn't need to know much about the app, what it must know are the various rules how to find the install location of .NET Core on the machine.

Reply:
> If we add an export to apphost it would only work for new apps, it won't fix your problem with existing apps.

We would have to keep the existing code for backcompat anyway, we do a lot more than the apphost does (traverse the PATH)

> I actually don't think you need nethost in every app - you need nethost in your code - to find hostfxr… has very little to do with the app. nethost has to be 100% backward compatible anyway, so it should work on old apps as well.

Not sure how I feel about redistributing nethost with AspNetCoreModule and if it brings any benefit.

Reply:
>Not sure how I feel about redistributing nethost with AspNetCoreModule and if it brings any benefit.
I think I'm confused. I though the general desire was to not have the "how to find hostfxr" logic in the ASP.NET code, but instead for that to come from some other component. I agree that in case of apps with `apphost` it can come from there, but apps which don't have `apphost` will still have that problem.

This API is also for cases where there's no app at all to begin with (for example the MSBuild/VS case of finding all available SDKs).